### ISSUE TITLE
[ML] Restarting a node while a data frame analytics job is running breaks MLMemoryTracker 

### ISSUE THREAD

**Issue by dolaru** [elastic/elasticsearch#44156]
Spotted in 7.3.0, single-node cluster.

**Problem**:
If a node is restarted while there are started data frame analytics jobs, `MlMemoryTracker` throws an exception because it's trying to access the `.ml-config` index too soon, and the index is not available for searching. Unfortunately, `MlMemoryTracker` never recovers from this.

As a result, anomaly detection jobs, datafeeds & data frame analytics jobs cannot be started/opened anymore. The ES APIs never return a response from the `_start`/`_open` requests.

**ES Log**:
```
[2019-07-09T18:27:08,442][WARN ][o.e.x.m.p.MlMemoryTracker] [dolaru-elastic-mbp.local] Failed to refresh job memory requirements
org.elasticsearch.action.search.SearchPhaseExecutionException: all shards failed
    at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:305) [elasticsearch-7.3.0.jar:7.3.0]
    at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:139) [elasticsearch-7.3.0.jar:7.3.0]
    at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:264) [elasticsearch-7.3.0.jar:7.3.0]
    at org.elasticsearch.action.search.InitialSearchPhase.onShardFailure(InitialSearchPhase.java:105) [elasticsearch-7.3.0.jar:7.3.0]
    at org.elasticsearch.action.search.InitialSearchPhase.lambda$performPhaseOnShard$1(InitialSearchPhase.java:251) [elasticsearch-7.3.0.jar:7.3.0]
    at org.elasticsearch.action.search.InitialSearchPhase$1.doRun(InitialSearchPhase.java:172) [elasticsearch-7.3.0.jar:7.3.0]
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-7.3.0.jar:7.3.0]
    at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:44) [elasticsearch-7.3.0.jar:7.3.0]
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:758) [elasticsearch-7.3.0.jar:7.3.0]
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-7.3.0.jar:7.3.0]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
    at java.lang.Thread.run(Thread.java:835) [?:?]
```

**Workaround**:
1. Force stop all data frame analytics jobs
2. Restart the node
3. `MlMemoryTracker` should throw anymore.

**Steps to reproduce**:
1. Create and start a data frame analytics job (or get one to fail - killing the `data_frame_analyzer` process should make it fail instantly).
2. Restart the node while the data frame analytics job task is still running.
3. Upon restart, during bootstrap, ES logs this:
```[2019-07-09T18:27:08,442][WARN ][o.e.x.m.p.MlMemoryTracker] [dolaru-elastic-mbp.local] Failed to refresh job memory requirements```
4. Notice that start/open doesn't work anymore for anomaly detection jobs, datafeeds & data frame analytics jobs

**Comment by droberts195** [comment]
There are two bits to this:

1. A theoretical bug that's existed since 6.6 that would cause the ML memory tracker to become permanently blocked if `onFailure` rather than `onResponse` was called on one of the lower level handlers.  However, there has been no known case of this being suffered by a user, not has it been seen in any testing.
2. The fact that in 7.3 we have added a path that is basically guaranteed to trigger the bug, rendering ML completely non-operational.

So it is the combination that makes this a blocker for 7.3.0.

I will fix it in two stages:

1. A fix to the problem of `onFailure` completely blocking up the ML memory tracker.  I'll backport this to 6.8 as theoretically it could be a problem in this version if some really obscure problem happens and it would be difficult to diagnose if a user reported it - the bug report would basically be "ML doesn't work at all", and that's time-consuming to tie down to a particular root cause.
2. An extra check in 7.3.0 and above that prevents the node assignment of data frame analytics jobs at node startup from triggering a failure of the ML memory tracker.  Even when the ML memory tracker recovers from failures it's best to avoid the log spam and confusion they'll cause.

**Comment by droberts195** [comment]
Fixed by a combination of #44158 and #44170.
